program/src/processor.rs

```diff
@@ -42,14 +45,17 @@ where
         return Err(SlashingError::ExceedsStatueOfLimitations.into());
     }
 
-    let (proof_data, context) = T::unpack(proof_data, instruction_data, accounts_info_iter)?;
+    let (proof_data, context) =
+        T::unpack_proof_and_context(proof_data, instruction_data, accounts_info_iter)?;
 
-    SlashingProofData::verify_proof(proof_data, context, slot, pubkey)?;
+    SlashingProofData::verify_proof(&proof_data, context, slot, &report.pubkey)?;
 
-    // TODO: follow up PR will record this violation in context state account. just
-    // log for now.
+    if !store_incident(slot, report, accounts, proof_data)? {
+        msg!("{} violation verified in slot {} however the violation has already been reported");
+        return Err(SlashingError::DuplicateReport.into());
```

Comment:
Unsure whether this should be an error or a success. Technically the transaction was valid, the reporter was just too slow.